keps/sig-etcd/4647-cluster-feature-gate/README.md

```diff
@@ -0,0 +1,482 @@
+
+
+### Risks and Mitigations
+
+#### Backport Risks
+
+Because we are adding a new type of raft request, if we want to support feature gate as soon as in 3.6, we have to backport the proto changes and add no-op apply logic in 3.5. 
+There is the potential risk of changing a stable release. But we do not think this is a real risk, because it should be a no-op change in 3.5. 
+When upgrading etcd cluster from 3.5 to 3.6, the users would need to upgrade to the latest 3.5 release first.
+
+If we do not backport to 3.5, we would need to add no-op changes in 3.6, and wait till 3.7 to add the real feature gate capabilities, which could be a long cycle of several years.
+
+#### Data Compatibility Risks during Feature Value Change
+
+There are some features that would need to touch data in the apply/storage layer. Disabling them might result in orphaned data or unrecognizable data that could be breaking.
+How do we handle the data change? 
+1. if the etcd server knows this feature, it should be able to handle the data when the feature is on or off. How it handles the data at WAL entry `i` is determined by the cluster feature flag set from raft before WAL entry `i`. 
+1. if the etcd server does not know this feature, i.e. its version is not compatible with this feature, it means the data is not compatible with the storage version. The storage version migration should handle cleaning up older or newly introduced data.
+
+Overall, when developers introduce new data with a feature, they should be careful to include data cleaning in `schema.UnsafeMigrate`. But we prefer not to have a mechanism to disallow such features to be turned off after they are turned on because new features could be buggy, we need to keep the option to be able to turn it off without bring down the whole cluster.
+
+#### Feature Implementation Change Risks
+
+There are cases when a developer wants to change how an existing feature works. This would make things complicated in a mixed version cluster. One member could have the old implementation while another member uses the latest implementation. How can we make sure the cluster works consistently?
+
+Ideally, the developer should introduced a new feature if they want to do major changes. If that is not desirable, similar to [how Kubernetes handles feature gating changes for compatibility version](https://github.com/kubernetes/enhancements/tree/master/keps/sig-architecture/4330-compatibility-versions#feature-gating-changes), we need to keep the implementation history for at least 2 minor versions (allowed version skew) in the code base if the feature change affects data in anyway, like
+```go
+if s.FeatureEnabled(FeatureA) && s.ClusterVersion() <= "3.6" {implementation 1}
+if s.FeatureEnabled(FeatureA) && s.ClusterVersion() >= "3.7" {implementation 2}
+```
+This way the cluster would work consistently because there is a single ClusterVersion across the whole cluster.
+
+## Design Details
```

Comment:
Please start from describing how features are enabled on command line. Does cluster feature gates share the same command line as server feature, I don't think they should. We should clarify this at the top.

Reply:
Also we should clarify one important thing, server level feature represent stability of the feature and can still have a separate enablement. For cluster level features we should focus on consistent enablement. Both share the same name "feature gate", but gating is different mechanism.